File: cpp/tests/unit-tests/SubscriptionManagerTest.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

#include <chrono>
#include <cstdint>
#include <memory>

#include <gtest/gtest.h>
#include <gmock/gmock.h>

#include "joynr/Future.h"
#include "joynr/ISubscriptionCallback.h"
#include "joynr/OnChangeSubscriptionQos.h"
#include "joynr/PeriodicSubscriptionQos.h"
#include "joynr/Runnable.h"
#include "joynr/SingleThreadedIOService.h"
#include "joynr/SubscriptionManager.h"
#include "joynr/UnicastSubscriptionCallback.h"
#include "joynr/Util.h"
#include "joynr/SubscriptionRequest.h"

#include "tests/utils/TimeUtils.h"
#include "tests/mock/MockDelayedScheduler.h"
#include "tests/mock/MockSubscriptionListener.h"

using ::testing::A;
using ::testing::_;
using ::testing::AtLeast;
using ::testing::AtMost;
using ::testing::Between;

MATCHER_P(publicationMissedException, subscriptionId, "")
{
    if (arg.getTypeName() == joynr::exceptions::PublicationMissedException::TYPE_NAME()) {
        joynr::exceptions::PublicationMissedException* errorArg =
                dynamic_cast<joynr::exceptions::PublicationMissedException*>(arg.clone());
        bool success = errorArg->getSubscriptionId() == subscriptionId &&
                       errorArg->getMessage() == subscriptionId;
        delete errorArg;
        return success;
    }
    return false;
}

using namespace joynr;

class SubscriptionManagerTest : public testing::Test
{
public:
    SubscriptionManagerTest() : singleThreadedIOService(std::make_shared<SingleThreadedIOService>())
    {
        singleThreadedIOService->start();
    }

    ~SubscriptionManagerTest()
    {
        singleThreadedIOService->stop();
    }

protected:
    std::shared_ptr<SingleThreadedIOService> singleThreadedIOService;
};

TEST_F(SubscriptionManagerTest, registerSubscription_subscriptionRequestIsCorrect)
{
    auto subscriptionManager =
            std::make_shared<SubscriptionManager>(singleThreadedIOService->getIOService(), nullptr);
    auto mockGpsSubscriptionListener =
            std::make_shared<MockSubscriptionListenerOneType<types::Localisation::GpsLocation>>();
    SubscriptionRequest subscriptionRequest;
    auto future = std::make_shared<Future<std::string>>();
    auto gpslocationCallback =
            std::make_shared<UnicastSubscriptionCallback<types::Localisation::GpsLocation>>(
                    subscriptionRequest.getSubscriptionId(), future, subscriptionManager);
    auto qos = std::make_shared<joynr::OnChangeSubscriptionQos>();
    std::int64_t now = static_cast<std::int64_t>(TimeUtils::getCurrentMillisSinceEpoch());
    qos->setExpiryDateMs(now + 10000);
    subscriptionManager->registerSubscription("methodName",
                                              gpslocationCallback,
                                              mockGpsSubscriptionListener,
                                              qos,
                                              subscriptionRequest);

    EXPECT_EQ("methodName", subscriptionRequest.getSubscribeToName());
    EXPECT_EQ(qos, subscriptionRequest.getQos());
}

TEST_F(SubscriptionManagerTest, registerSubscription_missedPublicationRunnableWorks)
{
    auto mockGpsSubscriptionListener =
            std::make_shared<MockSubscriptionListenerOneType<types::Localisation::GpsLocation>>();
    SubscriptionRequest subscriptionRequest;
    EXPECT_CALL(*mockGpsSubscriptionListener,
                onError(publicationMissedException(subscriptionRequest.getSubscriptionId())))
            .Times(AtLeast(4));
    auto future = std::make_shared<Future<std::string>>();
    auto subscriptionManager =
            std::make_shared<SubscriptionManager>(singleThreadedIOService->getIOService(), nullptr);
    auto gpslocationCallback =
            std::make_shared<UnicastSubscriptionCallback<types::Localisation::GpsLocation>>(
                    subscriptionRequest.getSubscriptionId(), future, subscriptionManager);
    auto qos = std::make_shared<joynr::PeriodicSubscriptionQos>(1100, 100, 100, 200);
    subscriptionManager->registerSubscription("methodName",
                                              gpslocationCallback,
                                              mockGpsSubscriptionListener,
                                              qos,
                                              subscriptionRequest);
    std::this_thread::sleep_for(std::chrono::milliseconds(1200));
}

TEST_F(SubscriptionManagerTest,
       registerSubscriptionWithSameSubscriptionId_missedPublicationRunnableWorks)
{
    auto mockGpsSubscriptionListener =
            std::make_shared<MockSubscriptionListenerOneType<types::Localisation::GpsLocation>>();
    SubscriptionRequest subscriptionRequest;
    EXPECT_CALL(*mockGpsSubscriptionListener,
                onError(publicationMissedException(subscriptionRequest.getSubscriptionId())))
            .Times(AtMost(6));
    auto future = std::make_shared<Future<std::string>>();
    auto subscriptionManager =
            std::make_shared<SubscriptionManager>(singleThreadedIOService->getIOService(), nullptr);
    auto gpslocationCallback =
            std::make_shared<UnicastSubscriptionCallback<types::Localisation::GpsLocation>>(
                    subscriptionRequest.getSubscriptionId(), future, subscriptionManager);
    auto qos = std::make_shared<PeriodicSubscriptionQos>(1100, 100, 100, 100);
    subscriptionManager->registerSubscription("methodName",
                                              gpslocationCallback,
                                              mockGpsSubscriptionListener,
                                              qos,
                                              subscriptionRequest);
    std::this_thread::sleep_for(std::chrono::milliseconds(300));
    auto mockGpsSubscriptionListener2 =
            std::make_shared<MockSubscriptionListenerOneType<types::Localisation::GpsLocation>>();
    EXPECT_CALL(*mockGpsSubscriptionListener2, onError(_)).Times(0);
    future = std::make_shared<Future<std::string>>();
    auto gpslocationCallback2 =
            std::make_shared<UnicastSubscriptionCallback<types::Localisation::GpsLocation>>(
                    subscriptionRequest.getSubscriptionId(), future, subscriptionManager);
    auto qos2 = std::make_shared<OnChangeSubscriptionQos>(700, 100, 100);
    subscriptionManager->registerSubscription("methodName",
                                              gpslocationCallback2,
                                              mockGpsSubscriptionListener2,
                                              qos2,
                                              subscriptionRequest);

    // now, no new publicationMissed callbacks are expected for the first subscriptionRequest
    std::this_thread::sleep_for(std::chrono::milliseconds(900));
}

TEST_F(SubscriptionManagerTest,
       registerSubscriptionWithSameSubscriptionId_correctDealingWithEnlargedExpiryDate)
{
    auto mockGpsSubscriptionListener =
            std::make_shared<MockSubscriptionListenerOneType<types::Localisation::GpsLocation>>();
    SubscriptionRequest subscriptionRequest;
    EXPECT_CALL(*mockGpsSubscriptionListener,
                onError(publicationMissedException(subscriptionRequest.getSubscriptionId())))
            .Times(AtLeast(6));
    auto future = std::make_shared<Future<std::string>>();
    auto subscriptionManager =
            std::make_shared<SubscriptionManager>(singleThreadedIOService->getIOService(), nullptr);
    auto gpslocationCallback =
            std::make_shared<UnicastSubscriptionCallback<types::Localisation::GpsLocation>>(
                    subscriptionRequest.getSubscriptionId(), future, subscriptionManager);
    auto qos = std::make_shared<PeriodicSubscriptionQos>(300, 100, 100, 100);
    subscriptionManager->registerSubscription("methodName",
                                              gpslocationCallback,
                                              mockGpsSubscriptionListener,
                                              qos,
                                              subscriptionRequest);

    qos = std::make_shared<PeriodicSubscriptionQos>(1000, 100, 100, 100);
    subscriptionManager->registerSubscription("methodName",
                                              gpslocationCallback,
                                              mockGpsSubscriptionListener,
                                              qos,
                                              subscriptionRequest);

    // now, no new publicationMissed callbacks are expected for the first subscriptionRequest
    std::this_thread::sleep_for(std::chrono::seconds(1));
}

TEST_F(SubscriptionManagerTest,
       registerSubscriptionWithSameSubscriptionId_correctDealingWithReducedExpiryDate)
{
    auto mockGpsSubscriptionListener =
            std::make_shared<MockSubscriptionListenerOneType<types::Localisation::GpsLocation>>();
    SubscriptionRequest subscriptionRequest;
    EXPECT_CALL(*mockGpsSubscriptionListener,
                onError(publicationMissedException(subscriptionRequest.getSubscriptionId())))
            .Times(AtMost(6));
    auto future = std::make_shared<Future<std::string>>();
    auto subscriptionManager =
            std::make_shared<SubscriptionManager>(singleThreadedIOService->getIOService(), nullptr);
    auto gpslocationCallback =
            std::make_shared<UnicastSubscriptionCallback<types::Localisation::GpsLocation>>(
                    subscriptionRequest.getSubscriptionId(), future, subscriptionManager);
    auto qos = std::make_shared<PeriodicSubscriptionQos>(1000, 100, 100, 100);
    subscriptionManager->registerSubscription("methodName",
                                              gpslocationCallback,
                                              mockGpsSubscriptionListener,
                                              qos,
                                              subscriptionRequest);

    qos = std::make_shared<PeriodicSubscriptionQos>(300, 100, 100, 100);
    subscriptionManager->registerSubscription("methodName",
                                              gpslocationCallback,
                                              mockGpsSubscriptionListener,
                                              qos,
                                              subscriptionRequest);

    // now, no new publicationMissed callbacks are expected for the first subscriptionRequest
    std::this_thread::sleep_for(std::chrono::seconds(1));
}

TEST_F(SubscriptionManagerTest, registerSubscription_withoutExpiryDate)
{
    auto mockGpsSubscriptionListener =
            std::make_shared<MockSubscriptionListenerOneType<types::Localisation::GpsLocation>>();
    auto future = std::make_shared<Future<std::string>>();
    auto mockDelayedScheduler =
            std::make_shared<MockDelayedScheduler>(singleThreadedIOService->getIOService());
    EXPECT_CALL(*mockDelayedScheduler, schedule(_, _)).Times(0);
    auto subscriptionManager = std::make_shared<SubscriptionManager>(mockDelayedScheduler, nullptr);
    SubscriptionRequest subscriptionRequest;
    auto gpslocationCallback =
            std::make_shared<UnicastSubscriptionCallback<types::Localisation::GpsLocation>>(
                    subscriptionRequest.getSubscriptionId(), future, subscriptionManager);
    auto qos = std::make_shared<OnChangeSubscriptionQos>(-1, 100, 100);
    subscriptionManager->registerSubscription("methodName",
                                              gpslocationCallback,
                                              mockGpsSubscriptionListener,
                                              qos,
                                              subscriptionRequest);
}

DelayedScheduler::RunnableHandle internalRunnableHandle = 1;

DelayedScheduler::RunnableHandle runnableHandle()
{
    return internalRunnableHandle++;
}

TEST_F(SubscriptionManagerTest, registerSubscription_withExpiryDate)
{
    auto mockGpsSubscriptionListener =
            std::make_shared<MockSubscriptionListenerOneType<types::Localisation::GpsLocation>>();
    auto future = std::make_shared<Future<std::string>>();
    auto mockDelayedScheduler =
            std::make_shared<MockDelayedScheduler>(singleThreadedIOService->getIOService());
    EXPECT_CALL(*mockDelayedScheduler, schedule(A<std::shared_ptr<Runnable>>(), _))
            .Times(1)
            .WillRepeatedly(::testing::Return(runnableHandle()));
    auto subscriptionManager = std::make_shared<SubscriptionManager>(mockDelayedScheduler, nullptr);
    SubscriptionRequest subscriptionRequest;
    auto gpslocationCallback =
            std::make_shared<UnicastSubscriptionCallback<types::Localisation::GpsLocation>>(
                    subscriptionRequest.getSubscriptionId(), future, subscriptionManager);
    auto qos = std::make_shared<OnChangeSubscriptionQos>(1000, 100, 100);
    subscriptionManager->registerSubscription("methodName",
                                              gpslocationCallback,
                                              mockGpsSubscriptionListener,
                                              qos,
                                              subscriptionRequest);
}

TEST_F(SubscriptionManagerTest,
       unregisterSubscription_unregisterLeadsToStoppingMissedPublicationRunnables)
{
    auto mockGpsSubscriptionListener =
            std::make_shared<MockSubscriptionListenerOneType<types::Localisation::GpsLocation>>();
    SubscriptionRequest subscriptionRequest;
    EXPECT_CALL(*mockGpsSubscriptionListener,
                onError(publicationMissedException(subscriptionRequest.getSubscriptionId())))
            .Times(Between(2, 3));
    auto subscriptionManager =
            std::make_shared<SubscriptionManager>(singleThreadedIOService->getIOService(), nullptr);
    auto future = std::make_shared<Future<std::string>>();
    auto gpslocationCallback =
            std::make_shared<UnicastSubscriptionCallback<types::Localisation::GpsLocation>>(
                    subscriptionRequest.getSubscriptionId(), future, subscriptionManager);
    auto qos = std::make_shared<PeriodicSubscriptionQos>(2000, // validity
                                                         100,  // publication ttl
                                                         100,  // period
                                                         400   // alert after interval
                                                         );
    subscriptionManager->registerSubscription("methodName",
                                              gpslocationCallback,
                                              mockGpsSubscriptionListener,
                                              qos,
                                              subscriptionRequest);
    std::this_thread::sleep_for(std::chrono::milliseconds(900));
    subscriptionManager->unregisterSubscription(subscriptionRequest.getSubscriptionId());
    std::this_thread::sleep_for(std::chrono::milliseconds(1100));
}

TEST_F(SubscriptionManagerTest, unregisterSubscription_unregisterLeadsOnNonExistantSubscription)
{
    auto subscriptionManager =
            std::make_shared<SubscriptionManager>(singleThreadedIOService->getIOService(), nullptr);
    subscriptionManager->unregisterSubscription("superId");
}

TEST_F(SubscriptionManagerTest, getSubscriptionListener)
{
    auto mockGpsSubscriptionListener =
            std::make_shared<MockSubscriptionListenerOneType<types::Localisation::GpsLocation>>();
    SubscriptionRequest subscriptionRequest;
    auto subscriptionManager =
            std::make_shared<SubscriptionManager>(singleThreadedIOService->getIOService(), nullptr);
    auto future = std::make_shared<Future<std::string>>();
    auto gpslocationCallback =
            std::make_shared<UnicastSubscriptionCallback<types::Localisation::GpsLocation>>(
                    subscriptionRequest.getSubscriptionId(), future, subscriptionManager);
    auto qos = std::make_shared<PeriodicSubscriptionQos>(2000, // validity
                                                         100,  // publication ttl
                                                         100,  // period
                                                         400   // alert after interval
                                                         );
    subscriptionManager->registerSubscription("broadcastName",
                                              gpslocationCallback,
                                              mockGpsSubscriptionListener,
                                              qos,
                                              subscriptionRequest);
    EXPECT_EQ(
            mockGpsSubscriptionListener,
            subscriptionManager->getSubscriptionListener(subscriptionRequest.getSubscriptionId()));
}

File: java/core/libjoynr/src/test/java/io/joynr/pubsub/subscription/SubscriptionManagerTest.java
package io.joynr.pubsub.subscription;

/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2015 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.times;
import io.joynr.dispatcher.Dispatcher;
import io.joynr.exceptions.JoynrMessageNotSentException;
import io.joynr.exceptions.JoynrSendBufferFullException;
import io.joynr.messaging.MessagingQos;
import io.joynr.proxy.invocation.AttributeSubscribeInvocation;
import io.joynr.proxy.invocation.BroadcastSubscribeInvocation;
import io.joynr.pubsub.PubSubState;
import io.joynr.pubsub.SubscriptionQos;

import java.io.IOException;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

import joynr.OnChangeSubscriptionQos;
import joynr.PeriodicSubscriptionQos;
import joynr.SubscriptionRequest;
import joynr.SubscriptionStop;
import joynr.tests.testBroadcastInterface.LocationUpdateBroadcastListener;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.runners.MockitoJUnitRunner;

import com.fasterxml.jackson.core.JsonGenerationException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.google.common.collect.Maps;

@RunWith(MockitoJUnitRunner.class)
public class SubscriptionManagerTest {

    private String attributeName;
    private AttributeSubscriptionListener<?> attributeSubscriptionCallback;

    private SubscriptionQos qos;
    private OnChangeSubscriptionQos onChangeQos;

    private SubscriptionQos qosWithoutExpiryDate;
    private MessagingQos qosSettings;

    @Mock
    ConcurrentMap<String, ScheduledFuture<?>> subscriptionEndFutures;

    private ConcurrentMap<String, AttributeSubscriptionListener<?>> attributeSubscriptionDirectory = spy(new ConcurrentHashMap<String, AttributeSubscriptionListener<?>>());
    private ConcurrentMap<String, BroadcastSubscriptionListener> broadcastSubscriptionDirectory = spy(new ConcurrentHashMap<String, BroadcastSubscriptionListener>());
    private ConcurrentMap<String, PubSubState> subscriptionStates = spy(new ConcurrentHashMap<String, PubSubState>());
    private ConcurrentMap<String, MissedPublicationTimer> missedPublicationTimers = spy(new ConcurrentHashMap<String, MissedPublicationTimer>());

    @Mock
    private PubSubState subscriptionState;

    private SubscriptionManager subscriptionManager;
    private String subscriptionId;
    private MissedPublicationTimer missedPublicationTimer;

    private String fromParticipantId;
    private String toParticipantId;

    @Mock
    private ConcurrentMap<String, Class<?>> subscriptionAttributeTypes;
    @Mock
    private ScheduledExecutorService cleanupScheduler;

    @Mock
    private Dispatcher dispatcher;

    @Before
    public void setUp() {
        subscriptionManager = new SubscriptionManagerImpl(attributeSubscriptionDirectory,
                                                          broadcastSubscriptionDirectory,
                                                          subscriptionStates,
                                                          missedPublicationTimers,
                                                          subscriptionEndFutures,
                                                          subscriptionAttributeTypes,
                                                          Maps.<String, Class<?>[]> newConcurrentMap(),
                                                          cleanupScheduler,
                                                          dispatcher);
        subscriptionId = "testSubscription";

        attributeName = "testAttribute";
        attributeSubscriptionCallback = new AttributeSubscriptionListener<Integer>() {
            @Override
            public void onError() {
                // TODO Auto-generated method stub
            }

            @Override
            public void onReceive(Integer value) {
                // TODO Auto-generated method stub

            }
        };
        long minInterval_ms = 100;
        long maxInterval_ms = 5000;
        long endDate_ms = System.currentTimeMillis() + 20000;
        long alertInterval_ms = 6000;
        long publicationTtl_ms = 1000;
        qos = new PeriodicSubscriptionQos(maxInterval_ms, endDate_ms, alertInterval_ms, publicationTtl_ms);

        onChangeQos = new OnChangeSubscriptionQos(minInterval_ms, endDate_ms, publicationTtl_ms);
        qosWithoutExpiryDate = new PeriodicSubscriptionQos(maxInterval_ms,
                                                           SubscriptionQos.NO_EXPIRY_DATE,
                                                           alertInterval_ms,
                                                           publicationTtl_ms);
        missedPublicationTimer = new MissedPublicationTimer(endDate_ms,
                                                            maxInterval_ms,
                                                            alertInterval_ms,
                                                            attributeSubscriptionCallback,
                                                            subscriptionState);

        qosSettings = new MessagingQos();
        fromParticipantId = "fromParticipantId";
        toParticipantId = "toParticipantId";
    }

    @Test
    public void registerSubscription() throws JoynrSendBufferFullException, JoynrMessageNotSentException,
                                      JsonGenerationException, JsonMappingException, IOException {
        class IntegerReference extends TypeReference<Integer> {
        }

        AttributeSubscribeInvocation subscriptionRequest = new AttributeSubscribeInvocation(attributeName,
                                                                                            IntegerReference.class,
                                                                                            attributeSubscriptionCallback,
                                                                                            qos,
                                                                                            null);
        subscriptionManager.registerAttributeSubscription(fromParticipantId, toParticipantId, subscriptionRequest);
        subscriptionId = subscriptionRequest.getSubscriptionId();

        Mockito.verify(attributeSubscriptionDirectory).put(Mockito.anyString(),
                                                           Mockito.eq(attributeSubscriptionCallback));
        Mockito.verify(subscriptionStates).put(Mockito.anyString(), Mockito.any(PubSubState.class));

        Mockito.verify(cleanupScheduler).schedule(Mockito.any(Runnable.class),
                                                  Mockito.eq(qos.getExpiryDate()),
                                                  Mockito.eq(TimeUnit.MILLISECONDS));
        Mockito.verify(subscriptionEndFutures, Mockito.times(1)).put(Mockito.eq(subscriptionId),
                                                                     Mockito.any(ScheduledFuture.class));

        Mockito.verify(dispatcher).sendSubscriptionRequest(eq(fromParticipantId),
                                                           eq(toParticipantId),
                                                           any(SubscriptionRequest.class),
                                                           any(MessagingQos.class),
                                                           eq(false));
    }

    @Test
    public void registerBroadcastSubscription() throws JoynrSendBufferFullException, JoynrMessageNotSentException,
                                               JsonGenerationException, JsonMappingException, IOException {
        String broadcastName = "broadcastName";
        BroadcastSubscriptionListener broadcastSubscriptionListener = mock(LocationUpdateBroadcastListener.class);
        BroadcastSubscribeInvocation subscriptionRequest = new BroadcastSubscribeInvocation(broadcastName,
                                                                                            broadcastSubscriptionListener,
                                                                                            onChangeQos,
                                                                                            null);
        subscriptionManager.registerBroadcastSubscription(fromParticipantId, toParticipantId, subscriptionRequest);
        subscriptionId = subscriptionRequest.getSubscriptionId();

        Mockito.verify(broadcastSubscriptionDirectory).put(Mockito.anyString(),
                                                           Mockito.eq(broadcastSubscriptionListener));
        Mockito.verify(subscriptionStates).put(Mockito.anyString(), Mockito.any(PubSubState.class));

        Mockito.verify(cleanupScheduler).schedule(Mockito.any(Runnable.class),
                                                  Mockito.eq(qos.getExpiryDate()),
                                                  Mockito.eq(TimeUnit.MILLISECONDS));
        Mockito.verify(subscriptionEndFutures, Mockito.times(1)).put(Mockito.eq(subscriptionId),
                                                                     Mockito.any(ScheduledFuture.class));

        Mockito.verify(dispatcher).sendSubscriptionRequest(eq(fromParticipantId),
                                                           eq(toParticipantId),
                                                           any(SubscriptionRequest.class),
                                                           any(MessagingQos.class),
                                                           eq(true));
    }

    @Test
    public void registerSubscriptionWithoutExpiryDate() throws JoynrSendBufferFullException,
                                                       JoynrMessageNotSentException, JsonGenerationException,
                                                       JsonMappingException, IOException {
        class IntegerReference extends TypeReference<Integer> {
        }

        AttributeSubscribeInvocation request = new AttributeSubscribeInvocation(attributeName,
                                                                                IntegerReference.class,
                                                                                attributeSubscriptionCallback,
                                                                                qosWithoutExpiryDate,
                                                                                null);
        subscriptionId = request.getSubscriptionId();
        subscriptionManager.registerAttributeSubscription(fromParticipantId, toParticipantId, request);

        Mockito.verify(attributeSubscriptionDirectory).put(Mockito.anyString(),
                                                           Mockito.eq(attributeSubscriptionCallback));
        Mockito.verify(subscriptionStates).put(Mockito.anyString(), Mockito.any(PubSubState.class));

        Mockito.verify(cleanupScheduler, never()).schedule(Mockito.any(Runnable.class),
                                                           Mockito.anyLong(),
                                                           Mockito.any(TimeUnit.class));
        Mockito.verify(subscriptionEndFutures, never()).put(Mockito.anyString(), Mockito.any(ScheduledFuture.class));

        Mockito.verify(dispatcher).sendSubscriptionRequest(eq(fromParticipantId),
                                                           eq(toParticipantId),
                                                           any(SubscriptionRequest.class),
                                                           any(MessagingQos.class),
                                                           eq(false));
    }

    @Test
    public void unregisterSubscription() throws JoynrSendBufferFullException, JoynrMessageNotSentException,
                                        JsonGenerationException, JsonMappingException, IOException {

        Mockito.when(subscriptionStates.get(subscriptionId)).thenReturn(subscriptionState);
        Mockito.when(missedPublicationTimers.containsKey(subscriptionId)).thenReturn(true);
        Mockito.when(missedPublicationTimers.get(subscriptionId)).thenReturn(missedPublicationTimer);
        subscriptionManager.unregisterSubscription(fromParticipantId, toParticipantId, subscriptionId, qosSettings);

        Mockito.verify(subscriptionStates).get(Mockito.eq(subscriptionId));
        Mockito.verify(subscriptionState).stop();

        Mockito.verify(dispatcher, times(1)).sendSubscriptionStop(Mockito.eq(fromParticipantId),
                                                                  Mockito.eq(toParticipantId),
                                                                  Mockito.eq(new SubscriptionStop(subscriptionId)),
                                                                  Mockito.any(MessagingQos.class));

    }
}
